beacon-chain/builder/service.go

```diff
@@ -154,12 +161,34 @@ func (s *Service) RegisterValidator(ctx context.Context, reg []*ethpb.SignedVali
 		idxs = append(idxs, nx)
 		msgs = append(msgs, r.Message)
 		valid = append(valid, r)
+		indexToRegistration[nx] = r.Message
 	}
 	if err := s.c.RegisterValidator(ctx, valid); err != nil {
 		return errors.Wrap(err, "could not register validator(s)")
 	}
 
-	return s.cfg.beaconDB.SaveRegistrationsByValidatorIDs(ctx, idxs, msgs)
+	if len(indexToRegistration) != len(msgs) {
+		return errors.New("ids and registrations must be the same length")
+	}
+	if s.registrationCache != nil {
+		s.registrationCache.UpdateIndexToRegisteredMap(indexToRegistration)
+		return nil
+	} else {
+		return s.cfg.beaconDB.SaveRegistrationsByValidatorIDs(ctx, idxs, msgs)
+	}
+}
+
+// RegistrationByValidatorID returns either the values from the cache or db.
+func (s *Service) RegistrationByValidatorID(ctx context.Context, id primitives.ValidatorIndex) (*ethpb.ValidatorRegistrationV1, error) {
+	if s.cfg == nil || s.cfg.beaconDB == nil {
+		return nil, errors.New("nil beacon db")
+	}
```

Comment:
Could this be a problem if you only had a registration cache but no db?

Reply:
perhaps the suggestion here is to move it below